storage/mysql/map_storage_test.go

```diff
@@ -32,64 +32,107 @@ func TestMySQLMapStorage_CheckDatabaseAccessible(t *testing.T) {
 	}
 }
 
-func TestMapBegin(t *testing.T) {
+func TestMapBeginSnapshot(t *testing.T) {
 	cleanTestDB(DB)
-	mapID := createMapForTests(DB)
-	storage := NewMapStorage(DB)
 
-	// TODO(codingllama): Add tree existence / type validation
+	frozenMapID := createMapForTests(DB)
+	updateTree(DB, frozenMapID, func(tree *trillian.Tree) {
+		tree.TreeState = trillian.TreeState_FROZEN
+	})
+
+	activeMapID := createMapForTests(DB)
+	logID := createLogForTests(DB)
+
 	tests := []struct {
+		desc  string
 		mapID int64
+		// "begin" and "snapshot" define which methods to test. Both may be specified for
+		// the same scenario.
+		begin, snapshot, wantErr bool
 	}{
-		{mapID: mapID},
+		{
+			desc:     "unknown",
+			mapID:    -1,
+			begin:    true,
+			snapshot: true,
+			wantErr:  true,
+		},
+		{
+			desc:     "activeMap",
+			mapID:    activeMapID,
+			begin:    true,
+			snapshot: true,
+		},
+		{
+			desc:    "frozenBegin",
+			mapID:   frozenMapID,
+			begin:   true,
+			wantErr: true,
+		},
+		{
+			desc:     "frozenSnapshot",
+			mapID:    frozenMapID,
+			snapshot: true,
+		},
+		{
+			desc:     "log",
+			mapID:    logID,
+			begin:    true,
+			snapshot: true,
+			wantErr:  true,
+		},
 	}
 
 	ctx := context.Background()
+	s := NewMapStorage(DB)
 	for _, test := range tests {
-		tx, err := storage.BeginForTree(ctx, test.mapID)
-		if err != nil {
-			t.Fatalf("Begin() = (_, %v), want = (_, nil)", err)
-		}
-		defer tx.Close()
-		root, err := tx.LatestSignedMapRoot()
-		if err != nil {
-			t.Errorf("LatestSignedMapRoot() = (_, %v), want = (_, nil)", err)
+		if !test.begin && !test.snapshot {
```

Comment:
This feels a little over-complicated for a test.  Would it be possible to simplify by doing exactly one of begin/snapshot per-test?  You could have a `writableTX bool` to indicate that, and just duplicate a few of the table entries; I think that would result in a more linear test structure with less function-pointer/lambda shenanigans.